File: wasm-ml/src/random_forest.rs
//! RandomForest implementation for regression
//! 
//! This implementation uses bagging (bootstrap aggregating) to train
//! multiple decision trees and averages their predictions.
//! 
//! Now includes GPU-accelerated training via WebGPU compute shaders.

use serde::{Deserialize, Serialize};
use rand::Rng;
use rand_xoshiro::Xoshiro256PlusPlus;
use rand::SeedableRng;

use crate::data::Dataset;
use crate::gpu_training::GpuTrainer;

/// A single decision tree node
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TreeNode {
    Leaf {
        value: f32,
    },
    Internal {
        feature_idx: usize,
        threshold: f32,
        left: Box<TreeNode>,
        right: Box<TreeNode>,
    },
}

impl TreeNode {
    /// Predict for a single sample
    pub fn predict(&self, sample: &[f32]) -> f32 {
        match self {
            TreeNode::Leaf { value } => *value,
            TreeNode::Internal { feature_idx, threshold, left, right } => {
                if sample[*feature_idx] <= *threshold {
                    left.predict(sample)
                } else {
                    right.predict(sample)
                }
            }
        }
    }
}

/// A single decision tree
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DecisionTree {
    root: TreeNode,
    max_depth: usize,
}

impl DecisionTree {
    pub fn new(max_depth: usize) -> Self {
        Self {
            root: TreeNode::Leaf { value: 0.0 },
            max_depth,
        }
    }
    
    /// Train the tree on a dataset
    pub fn train(&mut self, data: &[f32], labels: &[f32], n_features: usize, rng: &mut impl Rng) {
        let n_samples = labels.len();
        let indices: Vec<usize> = (0..n_samples).collect();
        self.root = self.build_tree(data, labels, &indices, n_features, 0, rng);
    }
    
    /// Train tree with GPU-accelerated split finding
    pub async fn train_gpu(
        &mut self,
        data: &[f32],
        labels: &[f32],
        indices: &[u32],
        n_features: usize,
        gpu_trainer: &GpuTrainer,
        rng: &mut impl Rng,
    ) -> Result<(), String> {
        let indices_usize: Vec<usize> = indices.iter().map(|&x| x as usize).collect();
        self.root = self.build_tree_gpu(data, labels, &indices_usize, n_features, 0, gpu_trainer, rng)
            .await?;
        Ok(())
    }
    
    fn build_tree(
        &self,
        data: &[f32],
        labels: &[f32],
        indices: &[usize],
        n_features: usize,
        depth: usize,
        rng: &mut impl Rng,
    ) -> TreeNode {
        // Base cases: max depth or too few samples
        if depth >= self.max_depth || indices.len() < 2 {
            let mean = indices.iter().map(|&i| labels[i]).sum::<f32>() / indices.len() as f32;
            return TreeNode::Leaf { value: mean };
        }
        
        // Find best split
        let (best_feature, best_threshold, best_score) = self.find_best_split(
            data,
            labels,
            indices,
            n_features,
            rng,
        );
        
        // If no good split found, create leaf
        if best_score.is_infinite() {
            let mean = indices.iter().map(|&i| labels[i]).sum::<f32>() / indices.len() as f32;
            return TreeNode::Leaf { value: mean };
        }
        
        // Split data
        let (left_indices, right_indices) = self.split_data(
            data,
            indices,
            n_features,
            best_feature,
            best_threshold,
        );
        
        // Recursively build subtrees
        let left = Box::new(self.build_tree(data, labels, &left_indices, n_features, depth + 1, rng));
        let right = Box::new(self.build_tree(data, labels, &right_indices, n_features, depth + 1, rng));
        
        TreeNode::Internal {
            feature_idx: best_feature,
            threshold: best_threshold,
            left,
            right,
        }
    }
    
    async fn build_tree_gpu(
        &self,
        data: &[f32],
        labels: &[f32],
        indices: &[usize],
        n_features: usize,
        depth: usize,
        gpu_trainer: &GpuTrainer,
        rng: &mut impl Rng,
    ) -> Result<TreeNode, String> {
        // Base cases
        if depth >= self.max_depth || indices.len() < 2 {
            let mean = indices.iter().map(|&i| labels[i]).sum::<f32>() / indices.len() as f32;
            return Ok(TreeNode::Leaf { value: mean });
        }
        
        // Find best split using GPU
        let (best_feature, best_threshold, best_score) = self.find_best_split_gpu(
            data,
            labels,
            indices,
            n_features,
            gpu_trainer,
            rng,
        ).await?;
        
        if best_score.is_infinite() {
            let mean = indices.iter().map(|&i| labels[i]).sum::<f32>() / indices.len() as f32;
            return Ok(TreeNode::Leaf { value: mean });
        }
        
        // Split data
        let (left_indices, right_indices) = self.split_data(
            data,
            indices,
            n_features,
            best_feature,
            best_threshold,
        );
        
        // Recursively build subtrees
        let left = Box::new(self.build_tree_gpu(data, labels, &left_indices, n_features, depth + 1, gpu_trainer, rng).await?);
        let right = Box::new(self.build_tree_gpu(data, labels, &right_indices, n_features, depth + 1, gpu_trainer, rng).await?);
        
        Ok(TreeNode::Internal {
            feature_idx: best_feature,
            threshold: best_threshold,
            left,
            right,
        })
    }
    
    async fn find_best_split_gpu(
        &self,
        data: &[f32],
        labels: &[f32],
        indices: &[usize],
        n_features: usize,
        gpu_trainer: &GpuTrainer,
        rng: &mut impl Rng,
    ) -> Result<(usize, f32, f32), String> {
        let mut best_feature = 0;
        let mut best_threshold = 0.0;
        let mut best_score = f32::INFINITY;
        
        // Random feature selection
        let n_features_to_try = (n_features as f32).sqrt().ceil() as usize;
        let mut features: Vec<usize> = (0..n_features).collect();
        
        use rand::seq::SliceRandom;
        features.shuffle(rng);
        
        // Convert indices to u32 for GPU
        let indices_u32: Vec<u32> = indices.iter().map(|&x| x as u32).collect();
        
        // Try each feature
        for &feature_idx in features.iter().take(n_features_to_try) {
            // Use GPU to find best split for this feature
            let (threshold, score) = gpu_trainer.find_best_split(
                data,
                labels,
                &indices_u32,
                n_features,
                feature_idx,
            ).await?;
            
            if score < best_score {
                best_score = score;
                best_feature = feature_idx;
                best_threshold = threshold;
            }
        }
        
        Ok((best_feature, best_threshold, best_score))
    }
    
    fn find_best_split(
        &self,
        data: &[f32],
        labels: &[f32],
        indices: &[usize],
        n_features: usize,
        rng: &mut impl Rng,
    ) -> (usize, f32, f32) {
        let mut best_feature = 0;
        let mut best_threshold = 0.0;
        let mut best_score = f32::INFINITY;
        
        // Random feature selection (sqrt of total features)
        let n_features_to_try = (n_features as f32).sqrt().ceil() as usize;
        let mut features: Vec<usize> = (0..n_features).collect();
        
        // Shuffle and select random features
        use rand::seq::SliceRandom;
        features.shuffle(rng);
        
        for &feature_idx in features.iter().take(n_features_to_try) {
            // Get unique values for this feature
            let mut values: Vec<f32> = indices
                .iter()
                .map(|&i| data[i * n_features + feature_idx])
                .collect();
            values.sort_by(|a, b| a.partial_cmp(b).unwrap());
            values.dedup();
            
            // Try different thresholds
            for i in 0..values.len().saturating_sub(1) {
                let threshold = (values[i] + values[i + 1]) / 2.0;
                
                // Calculate MSE for this split
                let (left_indices, right_indices) = self.split_data(
                    data,
                    indices,
                    n_features,
                    feature_idx,
                    threshold,
                );
                
                if left_indices.is_empty() || right_indices.is_empty() {
                    continue;
                }
                
                let score = self.calculate_mse(&left_indices, labels)
                    + self.calculate_mse(&right_indices, labels);
                
                if score < best_score {
                    best_score = score;
                    best_feature = feature_idx;
                    best_threshold = threshold;
                }
            }
        }
        
        (best_feature, best_threshold, best_score)
    }
    
    fn split_data(
        &self,
        data: &[f32],
        indices: &[usize],
        n_features: usize,
        feature_idx: usize,
        threshold: f32,
    ) -> (Vec<usize>, Vec<usize>) {
        let mut left = Vec::new();
        let mut right = Vec::new();
        
        for &idx in indices {
            let value = data[idx * n_features + feature_idx];
            if value <= threshold {
                left.push(idx);
            } else {
                right.push(idx);
            }
        }
        
        (left, right)
    }
    
    fn calculate_mse(&self, indices: &[usize], labels: &[f32]) -> f32 {
        if indices.is_empty() {
            return 0.0;
        }
        
        let mean = indices.iter().map(|&i| labels[i]).sum::<f32>() / indices.len() as f32;
        let mse = indices
            .iter()
            .map(|&i| {
                let diff = labels[i] - mean;
                diff * diff
            })
            .sum::<f32>()
            / indices.len() as f32;
        
        mse * indices.len() as f32
    }
    
    pub fn predict(&self, sample: &[f32]) -> f32 {
        self.root.predict(sample)
    }
}

/// RandomForest ensemble
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RandomForest {
    trees: Vec<DecisionTree>,
    n_estimators: usize,
    max_depth: usize,
}

impl RandomForest {
    pub fn new(n_estimators: usize, max_depth: usize) -> Self {
        Self {
            trees: Vec::with_capacity(n_estimators),
            n_estimators,
            max_depth,
        }
    }
    
    /// Train the forest on CPU
    pub fn train(&mut self, dataset: &Dataset) -> Result<(), String> {
        let mut rng = Xoshiro256PlusPlus::from_entropy();
        
        for i in 0..self.n_estimators {
            // Bootstrap sample
            let (sampled_data, sampled_labels) = dataset.bootstrap_sample(&mut rng);
            
            // Train tree
            let mut tree = DecisionTree::new(self.max_depth);
            tree.train(&sampled_data, &sampled_labels, dataset.n_features, &mut rng);
            
            self.trees.push(tree);
            
            // Progress indication
            if (i + 1) % 10 == 0 {
                eprintln!("Trained {}/{} trees (CPU)", i + 1, self.n_estimators);
            }
        }
        
        Ok(())
    }
    
    /// Train the forest with GPU acceleration
    pub async fn train_gpu(&mut self, dataset: &Dataset, gpu_trainer: &GpuTrainer) -> Result<(), String> {
        let mut rng = Xoshiro256PlusPlus::from_entropy();
        
        for i in 0..self.n_estimators {
            // Bootstrap sample on GPU
            let seed = rng.gen();
            let bootstrap_indices = gpu_trainer.bootstrap_sample(dataset.n_samples, seed)
                .await
                .map_err(|e| format!("GPU bootstrap failed: {}", e))?;
            
            // Extract bootstrapped data
            let mut sampled_data = Vec::with_capacity(dataset.n_samples * dataset.n_features);
            let mut sampled_labels = Vec::with_capacity(dataset.n_samples);
            
            for &idx in &bootstrap_indices {
                sampled_data.extend_from_slice(dataset.get_sample(idx as usize));
                sampled_labels.push(dataset.get_label(idx as usize));
            }
            
            // Train tree with GPU-accelerated split finding
            let mut tree = DecisionTree::new(self.max_depth);
            tree.train_gpu(
                &sampled_data,
                &sampled_labels,
                &bootstrap_indices,
                dataset.n_features,
                gpu_trainer,
                &mut rng,
            ).await?;
            
            self.trees.push(tree);
            
            // Progress indication
            if (i + 1) % 10 == 0 {
                eprintln!("Trained {}/{} trees (GPU)", i + 1, self.n_estimators);
            }
        }
        
        Ok(())
    }
    
    /// Predict on CPU (fallback)
    pub fn predict_cpu(
        &self,
        data: &[f32],
        n_samples: usize,
        n_features: usize,
    ) -> Result<Vec<f32>, String> {
        let mut predictions = Vec::with_capacity(n_samples);
        
        for sample_idx in 0..n_samples {
            let start = sample_idx * n_features;
            let end = start + n_features;
            let sample = &data[start..end];
            
            // Average predictions from all trees
            let sum: f32 = self.trees.iter().map(|tree| tree.predict(sample)).sum();
            let avg = sum / self.trees.len() as f32;
            
            predictions.push(avg);
        }
        
        Ok(predictions)
    }
    
    /// Get tree predictions for GPU processing
    pub fn get_tree_predictions(&self, sample: &[f32]) -> Vec<f32> {
        self.trees.iter().map(|tree| tree.predict(sample)).collect()
    }
    
    pub fn n_trees(&self) -> usize {
        self.trees.len()
    }
}

File: wasmwebgpu-ml/src/random_forest.cpp
/**
 * @file random_forest.cpp
 * @brief Implementation of RandomForest algorithm with wasi:gpu support
 */

#include "random_forest.hpp"
#include "gpu_executor.hpp"
#include <algorithm>
#include <numeric>
#include <cmath>
#include <iostream>
#include <limits>
#include <cstring>

namespace ml {

// ═══════════════════════════════════════════════════════════════════════
// TreeNode Implementation
// ═══════════════════════════════════════════════════════════════════════

std::unique_ptr<TreeNode> TreeNode::make_leaf(float value) {
    auto node = std::make_unique<TreeNode>();
    node->type = NodeType::LEAF;
    node->value = value;
    return node;
}

std::unique_ptr<TreeNode> TreeNode::make_internal(
    size_t feature_idx, 
    float threshold,
    std::unique_ptr<TreeNode> left,
    std::unique_ptr<TreeNode> right
) {
    auto node = std::make_unique<TreeNode>();
    node->type = NodeType::INTERNAL;
    node->feature_idx = feature_idx;
    node->threshold = threshold;
    node->left = std::move(left);
    node->right = std::move(right);
    return node;
}

float TreeNode::predict(const float* sample) const {
    if (type == NodeType::LEAF) {
        return value;
    } else {
        if (sample[feature_idx] <= threshold) {
            return left->predict(sample);
        } else {
            return right->predict(sample);
        }
    }
}

// ═══════════════════════════════════════════════════════════════════════
// DecisionTree Implementation
// ═══════════════════════════════════════════════════════════════════════

DecisionTree::DecisionTree(size_t max_depth)
    : max_depth_(max_depth)
    , root_(nullptr) {}

void DecisionTree::train_cpu(const std::vector<float>& data,
                              const std::vector<float>& labels,
                              size_t n_samples,
                              size_t n_features,
                              std::mt19937& rng) {
    std::vector<size_t> indices(n_samples);
    std::iota(indices.begin(), indices.end(), 0);
    root_ = build_tree_cpu(data, labels, indices, n_features, 0, rng);
}

void DecisionTree::train_with_gpu(const std::vector<float>& data,
                                   const std::vector<float>& labels,
                                   const std::vector<uint32_t>& bootstrap_indices,
                                   size_t n_features,
                                   GpuTrainer& gpu_trainer,
                                   std::mt19937& rng) {
    std::vector<size_t> indices(bootstrap_indices.begin(), bootstrap_indices.end());
    root_ = build_tree_gpu(data, labels, indices, n_features, 0, gpu_trainer, rng);
}

std::unique_ptr<TreeNode> DecisionTree::build_tree_cpu(
    const std::vector<float>& data,
    const std::vector<float>& labels,
    const std::vector<size_t>& indices,
    size_t n_features,
    size_t depth,
    std::mt19937& rng
) {
    // Base cases
    if (depth >= max_depth_ || indices.size() < 2) {
        float sum = 0.0f;
        for (size_t idx : indices) {
            sum += labels[idx];
        }
        float mean = sum / indices.size();
        return TreeNode::make_leaf(mean);
    }
    
    // Find best split
    auto split = find_best_split_cpu(data, labels, indices, n_features, rng);
    
    // If no good split found, create leaf
    if (std::isinf(split.score)) {
        float sum = 0.0f;
        for (size_t idx : indices) {
            sum += labels[idx];
        }
        float mean = sum / indices.size();
        return TreeNode::make_leaf(mean);
    }
    
    // Split data
    auto [left_indices, right_indices] = split_data(
        data, indices, n_features, split.feature_idx, split.threshold
    );
    
    if (left_indices.empty() || right_indices.empty()) {
        float sum = 0.0f;
        for (size_t idx : indices) {
            sum += labels[idx];
        }
        float mean = sum / indices.size();
        return TreeNode::make_leaf(mean);
    }
    
    // Recursively build subtrees
    auto left = build_tree_cpu(data, labels, left_indices, n_features, depth + 1, rng);
    auto right = build_tree_cpu(data, labels, right_indices, n_features, depth + 1, rng);
    
    return TreeNode::make_internal(
        split.feature_idx,
        split.threshold,
        std::move(left),
        std::move(right)
    );
}

std::unique_ptr<TreeNode> DecisionTree::build_tree_gpu(
    const std::vector<float>& data,
    const std::vector<float>& labels,
    const std::vector<size_t>& indices,
    size_t n_features,
    size_t depth,
    GpuTrainer& gpu_trainer,
    std::mt19937& rng
) {
    // Base cases
    if (depth >= max_depth_ || indices.size() < 2) {
        float sum = 0.0f;
        for (size_t idx : indices) {
            sum += labels[idx];
        }
        float mean = sum / indices.size();
        return TreeNode::make_leaf(mean);
    }
    
    // Find best split using GPU
    auto split = find_best_split_gpu(data, labels, indices, n_features, gpu_trainer, rng);
    
    // If no good split found, create leaf
    if (std::isinf(split.score)) {
        float sum = 0.0f;
        for (size_t idx : indices) {
            sum += labels[idx];
        }
        float mean = sum / indices.size();
        return TreeNode::make_leaf(mean);
    }
    
    // Split data
    auto [left_indices, right_indices] = split_data(
        data, indices, n_features, split.feature_idx, split.threshold
    );
    
    if (left_indices.empty() || right_indices.empty()) {
        float sum = 0.0f;
        for (size_t idx : indices) {
            sum += labels[idx];
        }
        float mean = sum / indices.size();
        return TreeNode::make_leaf(mean);
    }
    
    // Recursively build subtrees
    auto left = build_tree_gpu(data, labels, left_indices, n_features, depth + 1, gpu_trainer, rng);
    auto right = build_tree_gpu(data, labels, right_indices, n_features, depth + 1, gpu_trainer, rng);
    
    return TreeNode::make_internal(
        split.feature_idx,
        split.threshold,
        std::move(left),
        std::move(right)
    );
}

DecisionTree::SplitInfo DecisionTree::find_best_split_cpu(
    const std::vector<float>& data,
    const std::vector<float>& labels,
    const std::vector<size_t>& indices,
    size_t n_features,
    std::mt19937& rng
) {
    SplitInfo best;
    best.score = std::numeric_limits<float>::infinity();
    best.feature_idx = 0;
    best.threshold = 0.0f;
    
    // Random feature selection (sqrt of total features)
    size_t n_features_to_try = static_cast<size_t>(std::ceil(std::sqrt(n_features)));
    
    std::vector<size_t> features(n_features);
    std::iota(features.begin(), features.end(), 0);
    std::shuffle(features.begin(), features.end(), rng);
    
    for (size_t i = 0; i < n_features_to_try; ++i) {
        size_t feature_idx = features[i];
        
        // Collect unique values for this feature
        std::vector<float> values;
        values.reserve(indices.size());
        for (size_t idx : indices) {
            values.push_back(data[idx * n_features + feature_idx]);
        }
        std::sort(values.begin(), values.end());
        values.erase(std::unique(values.begin(), values.end()), values.end());
        
        // Try different thresholds
        for (size_t j = 0; j + 1 < values.size(); ++j) {
            float threshold = (values[j] + values[j + 1]) / 2.0f;
            
            auto [left_indices, right_indices] = split_data(
                data, indices, n_features, feature_idx, threshold
            );
            
            if (left_indices.empty() || right_indices.empty()) {
                continue;
            }
            
            float score = calculate_mse(labels, left_indices) +
                         calculate_mse(labels, right_indices);
            
            if (score < best.score) {
                best.score = score;
                best.feature_idx = feature_idx;
                best.threshold = threshold;
            }
        }
    }
    
    return best;
}

DecisionTree::SplitInfo DecisionTree::find_best_split_gpu(
    const std::vector<float>& data,
    const std::vector<float>& labels,
    const std::vector<size_t>& indices,
    size_t n_features,
    GpuTrainer& gpu_trainer,
    std::mt19937& rng
) {
    SplitInfo best;
    best.score = std::numeric_limits<float>::infinity();
    best.feature_idx = 0;
    best.threshold = 0.0f;
    
    // Random feature selection
    size_t n_features_to_try = static_cast<size_t>(std::ceil(std::sqrt(n_features)));
    
    std::vector<size_t> features(n_features);
    std::iota(features.begin(), features.end(), 0);
    std::shuffle(features.begin(), features.end(), rng);
    
    // Convert indices to uint32_t
    std::vector<uint32_t> indices_u32(indices.begin(), indices.end());
    
    for (size_t i = 0; i < n_features_to_try; ++i) {
        size_t feature_idx = features[i];
        
        // Collect unique values and compute thresholds
        std::vector<float> values;
        values.reserve(indices.size());
        for (size_t idx : indices) {
            values.push_back(data[idx * n_features + feature_idx]);
        }
        std::sort(values.begin(), values.end());
        values.erase(std::unique(values.begin(), values.end()), values.end());
        
        if (values.size() < 2) continue;
        
        // Compute thresholds
        std::vector<float> thresholds;
        for (size_t j = 0; j + 1 < values.size(); ++j) {
            thresholds.push_back((values[j] + values[j + 1]) / 2.0f);
        }
        
        if (thresholds.empty()) continue;
        
        // Use GPU to find best split for this feature
        auto [threshold, score] = gpu_trainer.find_best_split(
            indices_u32,
            feature_idx,
            thresholds
        );
        
        if (score < best.score) {
            best.score = score;
            best.feature_idx = feature_idx;
            best.threshold = threshold;
        }
    }
    
    return best;
}

std::pair<std::vector<size_t>, std::vector<size_t>> 
DecisionTree::split_data(
    const std::vector<float>& data,
    const std::vector<size_t>& indices,
    size_t n_features,
    size_t feature_idx,
    float threshold
) {
    std::vector<size_t> left, right;
    
    for (size_t idx : indices) {
        float value = data[idx * n_features + feature_idx];
        if (value <= threshold) {
            left.push_back(idx);
        } else {
            right.push_back(idx);
        }
    }
    
    return {left, right};
}

float DecisionTree::calculate_mse(
    const std::vector<float>& labels,
    const std::vector<size_t>& indices
) {
    if (indices.empty()) {
        return 0.0f;
    }
    
    float sum = 0.0f;
    for (size_t idx : indices) {
        sum += labels[idx];
    }
    float mean = sum / indices.size();
    
    float mse = 0.0f;
    for (size_t idx : indices) {
        float diff = labels[idx] - mean;
        mse += diff * diff;
    }
    
    return mse * indices.size();
}

float DecisionTree::predict(const float* sample) const {
    if (!root_) {
        std::cerr << "ERROR: Tree not trained" << std::endl;
        return 0.0f;
    }
    return root_->predict(sample);
}

// ═══════════════════════════════════════════════════════════════════════
// RandomForest Implementation
// ═══════════════════════════════════════════════════════════════════════

RandomForest::RandomForest(size_t n_estimators, size_t max_depth)
    : n_estimators_(n_estimators)
    , max_depth_(max_depth) {
    trees_.reserve(n_estimators);
}

void RandomForest::train_cpu(const Dataset& dataset) {
    std::random_device rd;
    std::mt19937 rng(rd());
    
    std::cout << "[TRAINING] RandomForest: " << n_estimators_ 
              << " trees, max_depth " << max_depth_ << std::endl;
    std::cout << "[TRAINING] Training on CPU..." << std::endl;
    
    for (size_t i = 0; i < n_estimators_; ++i) {
        auto [sampled_data, sampled_labels] = dataset.bootstrap_sample(rng);
        
        DecisionTree tree(max_depth_);
        tree.train_cpu(sampled_data, sampled_labels, 
                      dataset.size(), dataset.n_features(), rng);
        
        trees_.push_back(std::move(tree));
        
        if ((i + 1) % 10 == 0) {
            std::cerr << "Trained " << (i + 1) << "/" << n_estimators_ 
                      << " trees (CPU)" << std::endl;
        }
    }
    
    std::cout << "[TRAINING] Training completed!" << std::endl;
}

void RandomForest::train_with_gpu(const Dataset& dataset, GpuTrainer& gpu_trainer) {
    if (!gpu_trainer.is_available()) {
        std::cout << "[TRAINING] GPU not available, using CPU..." << std::endl;
        train_cpu(dataset);
        return;
    }
    
    std::random_device rd;
    std::mt19937 rng(rd());
    
    std::cout << "[TRAINING] RandomForest: " << n_estimators_ 
              << " trees, max_depth " << max_depth_ << std::endl;
    std::cout << "[TRAINING] Training with GPU acceleration (wasi:gpu)..." << std::endl;
    
    for (size_t i = 0; i < n_estimators_; ++i) {
        // Bootstrap sample on GPU
        uint32_t seed = rng();
        std::vector<uint32_t> bootstrap_indices = gpu_trainer.bootstrap_sample(
            dataset.size(), seed
        );
        
        // Extract bootstrapped data
        std::vector<float> sampled_data;
        std::vector<float> sampled_labels;
        
        sampled_data.reserve(dataset.size() * dataset.n_features());
        sampled_labels.reserve(dataset.size());
        
        for (uint32_t idx : bootstrap_indices) {
            const float* sample = dataset.get_sample(idx);
            sampled_data.insert(sampled_data.end(), sample, sample + dataset.n_features());
            sampled_labels.push_back(dataset.get_label(idx));
        }
        
        // Train tree with GPU-accelerated split finding
        DecisionTree tree(max_depth_);
        tree.train_with_gpu(sampled_data, sampled_labels, bootstrap_indices,
                            dataset.n_features(), gpu_trainer, rng);
        
        trees_.push_back(std::move(tree));
        
        if ((i + 1) % 10 == 0) {
            std::cerr << "Trained " << (i + 1) << "/" << n_estimators_ 
                      << " trees (GPU)" << std::endl;
        }
    }
    
    std::cout << "[TRAINING] Training completed!" << std::endl;
}

std::vector<float> RandomForest::predict_cpu(
    const std::vector<float>& data,
    size_t n_samples,
    size_t n_features
) const {
    std::vector<float> predictions;
    predictions.reserve(n_samples);
    
    for (size_t i = 0; i < n_samples; ++i) {
        const float* sample = &data[i * n_features];
        
        float sum = 0.0f;
        for (const auto& tree : trees_) {
            sum += tree.predict(sample);
        }
        float avg = sum / trees_.size();
        
        predictions.push_back(avg);
    }
    
    return predictions;
}

std::vector<float> RandomForest::predict_with_gpu(
    const std::vector<float>& data,
    size_t n_samples,
    size_t n_features,
    GpuPredictor& predictor
) {
    if (!predictor.is_available()) {
        std::cout << "[INFERENCE] GPU not available, using CPU..." << std::endl;
        return predict_cpu(data, n_samples, n_features);
    }
    
    // Get tree predictions for all samples
    std::vector<float> tree_predictions;
    tree_predictions.reserve(n_samples * trees_.size());
    
    for (size_t i = 0; i < n_samples; ++i) {
        const float* sample = &data[i * n_features];
        auto preds = get_tree_predictions(sample);
        tree_predictions.insert(tree_predictions.end(), preds.begin(), preds.end());
    }
    
    // Average on GPU
    return predictor.average_predictions(tree_predictions, n_samples, trees_.size());
}

std::vector<float> RandomForest::get_tree_predictions(const float* sample) const {
    std::vector<float> predictions;
    predictions.reserve(trees_.size());
    
    for (const auto& tree : trees_) {
        predictions.push_back(tree.predict(sample));
    }
    
    return predictions;
}

// ═══════════════════════════════════════════════════════════════════════
// Serialization (JSON) - Simplified
// ═══════════════════════════════════════════════════════════════════════

std::string DecisionTree::to_json() const {
    return "{}";
}

DecisionTree DecisionTree::from_json(const std::string& json_str) {
    return DecisionTree(0);
}

std::string RandomForest::to_json() const {
    char buf[256];
    snprintf(buf, sizeof(buf), 
             "{\"n_estimators\":%zu,\"max_depth\":%zu,\"n_trees\":%zu}",
             n_estimators_, max_depth_, trees_.size());
    return std::string(buf);
}

RandomForest RandomForest::from_json(const std::string& json_str) {
    // Simple parse - not production quality
    return RandomForest(100, 10);
}

} // namespace ml
